File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} subRoot
 * @return {boolean}
 */
var isSubtree = function (root, subRoot) {
    let queue1 = [root]
    let queue2 = [subRoot]
    let isvalid = false
    while (queue1.length > 0) {
        let length = queue1.length
        for (let i = 0; i < length; i++) {
            let first = queue1.shift()
            if (first.val != subRoot.val) {

                if (first.left) {
                    queue1.push(first.left)
                }
                if (first.right) {
                    queue1.push(first.right)
                }
            }
            else if (first.val == subRoot.val) {
                let ans = check(first, subRoot)
                if (ans) {
                    return isvalid = true
                }
                if (first.left) queue1.push(first.left);
                if (first.right) queue1.push(first.right);
            }
        }
    }
    return isvalid
};

var check = function (node1, node2) {
    if (!node1 && !node2) return true;
    if (!node1 || !node2) return false;
    if (node1.val !== node2.val) return false;

    return check(node1.left, node2.left) && check(node1.right, node2.right);
};

File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not subRoot:
            return True
        if not root:
            return False
        
        queue = deque([root])
        while queue:
            current = queue.popleft()
            if current and current.val == subRoot.val and self.sametree(current, subRoot):
                return True
            if current:
                queue.append(current.left)
                queue.append(current.right)
        return False

    def sametree(self, root, subRoot):
        if not root and not subRoot:
            return True
        if root and subRoot and root.val == subRoot.val:
            return self.sametree(root.left, subRoot.left) and self.sametree(
                root.right, subRoot.right
            )
        return False